src/enum.ts

```diff
@@ -0,0 +1,73 @@
+import { basicValidation, Check, Coerce, Convert, createAsCheck, createIsCheck, createMaybeAsCheck, createMaybeCheck, Validate } from './common';
+import { Issue, ValueProcessor } from './types';
+
+type EnumValue = string | number;
+type EnumType = { [key: string]: EnumValue };
+
+interface ValidationOptions { }
+
+class Generic<T> {
+  private keyToValue: { [key: string]: T } = {};
+  private valueToKey: { [key: string]: string } = {};
+
+  constructor(type: EnumType) {
+    for (const key in type) {
+      if (key) {
+        if (!/^\d+$/gi.test(key)) {
+          this.keyToValue[key] = type[key] as unknown as T;
+          this.valueToKey[type[key]] = key;
+        }
+      }
+    }
+  }
+
+  public check: (expectInEnum?: boolean) => Check<T> = (expectInEnum = false) => {
+    return (value): value is T => {
+      return expectInEnum
+        ? (value as EnumValue) in this.valueToKey
+        : (typeof value === 'string' || typeof value === 'number');
+    };
+  };
+
+  public coerce: Coerce<any, any> = () => (next) => (value, path) => {
+    return next(value, path);
+  };
+
+  public convert: Convert<T> = (value) => {
+    if ((value as EnumValue) in this.keyToValue) {
+      return this.keyToValue[value as EnumValue];
+    }
+    return undefined;
+  };
+
+  public validate: Validate<T, ValidationOptions> = (value, path, options) => {
+    const result = basicValidation(value, path, options);
+    if (!(value in this.valueToKey)) {
+      result.issues.push(Issue.forPath(path, value, 'key-not-found', {
+        validKeys: Object.values(this.keyToValue),
+        validValues: Object.keys(this.keyToValue),
```

Comment:
Nice idea to provide these as it might be useful for developers!

However we might consider if this can leak some unintended information to hostile users of our API. I would recommend to keep this, but as an opt-in feature in options, @gregbacchus?

Reply:
Ah yes, I like having this an opt-in feature. You're right that developers may not want this functionality 👍🏻. I'll get a commit up in a bit with this improvment

Reply:
Ok, I've got this enhancement added!